src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py

```diff
@@ -118,3 +120,9 @@ def __init__(
         self.time_stride = time_stride
         self.max_length = max_length
         self.num_mel_bins = num_mel_bins
+
+    # Overwritten from the parent class: AST is not compatible with `generate`, but has a config parameter sharing the
+    # same name (`max_length`). Sharing the same name triggers checks regarding the config -> generation_config
+    # generative parameters deprecation cycle, overwriting this function prevents this from happening.
+    def _get_non_default_generation_parameters(self) -> Dict[str, Any]:
+        return {}
```

Comment:
I'm not sure I understand this - why would there be checks run on a config -> generation_config if the model isn't compatible with generate? 

It seems like a bad pattern to need a generate specific method added to config if the model can't be used for generation

Reply:
I understand and share your pain 😬 Allow me to elaborate.

The root issue is in the model config. A model config is agnostic to whether the model can or cannot generate, unless we hardcode it in the config class or access external objects. 

When we call `config.save_pretrained`, we don't know whether we should or should not check generation parameters. Since we want to move all generation-related logic to `GenerationConfig`, whose one of the goals is to avoid clashes like this and allow proper validation, we have to assume all configs may be holding generate parameters. The current logic enforces no custom generate parameters are held in the model config, so no new models parameterize `generate` through the model config.

AST is the exception. It doesn't support `generate`, but it has a model parameter whose name is also used in `generate` (`max_length`). If a user creates a model with a custom `max_length`, which should be allowed, the current `save_pretrained` would crash because it thinks it has a custom `generate` parameter. This workaround effectively skips the `generate` checks, since we know it can't generate.

Between all three alternatives I could see, this seemed the best solution -- especially since it is a temporary fix, to be removed after we untangle the two configs (v5.0?)

Considered fixes:
- skip generate-related checks on this specific class (implemented)
- add a parameter to ALL model config classes to store whether the model can generate
- add an argument to `config.save_pretrained` to check whether to check `generate` parameters, receive `model.can_generate()` when called from `model.save_pretrained`

Reply:
Make sense - thanks for taking the time to explain!